rotors_gazebo_plugins/src/gazebo_wind_plugin.cpp

```diff
@@ -99,38 +115,68 @@ void GazeboWindPlugin::OnUpdate(const common::UpdateInfo& _info) {
-  wrench_msg.wrench.force.y = wind.y + wind_gust.y;
-  wrench_msg.wrench.force.z = wind.z + wind_gust.z;
-  wrench_msg.wrench.torque.x = 0;
-  wrench_msg.wrench.torque.y = 0;
-  wrench_msg.wrench.torque.z = 0;
+    geometry_msgs::WrenchStamped wrench_msg;
 
-  wind_pub_.publish(wrench_msg);
+    wrench_msg.header.frame_id = frame_id_;
+    wrench_msg.header.stamp.sec = now.sec;
+    wrench_msg.header.stamp.nsec = now.nsec;
+    wrench_msg.wrench.force.x = wind.x + wind_gust.x;
+    wrench_msg.wrench.force.y = wind.y + wind_gust.y;
+    wrench_msg.wrench.force.z = wind.z + wind_gust.z;
+    wrench_msg.wrench.torque.x = 0;
+    wrench_msg.wrench.torque.y = 0;
+    wrench_msg.wrench.torque.z = 0;
 
-  // Calculate the wind speed
-  double wind_speed = wind_speed_mean_; // + wind_speed_n_(random_generator_);
-  math::Vector3 wind_velocity = wind_speed * wind_direction_;
+    wind_pub_.publish(wrench_msg);
+
+    // Calculate the wind speed
+    double wind_speed = wind_speed_mean_; // + wind_speed_n_(random_generator_);
+    wind_velocity = wind_speed * wind_direction_;
+  } else {
+    // Get the current position of the aircraft in world coordinates
+    math::Vector3 link_position = link_->GetWorldPose().pos;
+
+    // Find indices for x,y and minimal/maximal values of the vertical spacing factor of the four surrounding grid columns
+    int i_inf = floor((link_position.x - min_x_) / res_x_);
+    int i_sup = i_inf + 1;
+    int j_inf = floor((link_position.y - min_y_) / res_y_);
+    int j_sup = j_inf + 1;
+
+    float vertical_spacing_factor_0 = (link_position.z - bottom_z_[i_inf + (j_inf * n_x_)]) / (top_z_[i_inf + (j_inf * n_x_)] - bottom_z_[i_inf + (j_inf * n_x_)]);
+    float vertical_spacing_factor_1 = (link_position.z - bottom_z_[i_sup + (j_inf * n_x_)]) / (top_z_[i_sup + (j_inf * n_x_)] - bottom_z_[i_sup + (j_inf * n_x_)]);
+    float vertical_spacing_factor_4 = (link_position.z - bottom_z_[i_inf + (j_sup * n_x_)]) / (top_z_[i_inf + (j_sup * n_x_)] - bottom_z_[i_inf + (j_sup * n_x_)]);
+    float vertical_spacing_factor_5 = (link_position.z - bottom_z_[i_sup + (j_sup * n_x_)]) / (top_z_[i_sup + (j_sup * n_x_)] - bottom_z_[i_sup + (j_sup * n_x_)]);
+
+    float vertical_spacing_factor_min = std::min(std::min(std::min(vertical_spacing_factor_0,vertical_spacing_factor_1),vertical_spacing_factor_4),vertical_spacing_factor_5);
+    float vertical_spacing_factor_max = std::max(std::max(std::max(vertical_spacing_factor_0,vertical_spacing_factor_1),vertical_spacing_factor_4),vertical_spacing_factor_5);
+
+    // Check if aircraft is out of wind field or not, and act accordingly
+    if (!( i_inf < 0 || j_inf < 0 || vertical_spacing_factor_max < 0 || i_sup > (n_x_ - 1) || j_sup > (n_y_ - 1) || vertical_spacing_factor_min > 1 )) {
+      InterpolateWindVelocity(link_position, i_inf, i_sup, j_inf, j_sup, vertical_spacing_factor_0, vertical_spacing_factor_1, vertical_spacing_factor_4, vertical_spacing_factor_5, wind_velocity);
```

Comment:
I found the approach very neat indeed! I have extended it for it to work in the wind plugin. I have also modified the cell vertices numbering once more so that most of the code could be simplified with loops and arrays while staying consistent.